source/host/KReplaySystem.cc

```diff
@@ -0,0 +1,209 @@
+#include "KReplaySystem.hh"
+
+#include "host/Option.hh"
+#include "host/SceneCreatorDynamic.hh"
+
+#include <abstract/File.hh>
+
+#include <game/system/RaceManager.hh>
+
+/// @brief Initializes the system.
+void KReplaySystem::init() {
+    ASSERT(m_currentGhostFileName);
+    ASSERT(m_currentRawGhost);
+    ASSERT(m_currentGhost);
+
+    auto *sceneCreator = new Host::SceneCreatorDynamic;
+    m_sceneMgr = new EGG::SceneManager(sceneCreator);
+
+    System::RaceConfig::RegisterInitCallback(OnInit, nullptr);
+    Abstract::File::Remove("results.txt");
+
+    m_sceneMgr->changeScene(0);
+}
+
+/// @brief Executes a frame.
+void KReplaySystem::calc() {
+    m_sceneMgr->calc();
+}
+
+/// @brief Executes a run.
+/// @details A run consists of replaying a ghost.
+/// @return Whether the run was successful or not.
+bool KReplaySystem::run() {
+    while (!calcEnd()) {
+        calc();
+    }
+
+    return success();
+}
+
+/// @brief Parses non-generic command line options.
+/// @details The only currently accepted option is the ghost flag.
+/// @param argc The number of arguments.
+/// @param argv The arguments.
+void KReplaySystem::parseOptions(int argc, char **argv) {
+    if (argc < 2) {
+        PANIC("Expected ghost argument!");
+    }
+
+    for (int i = 0; i < argc; ++i) {
+        std::optional<Host::EOption> flag = Host::Option::CheckFlag(argv[i]);
+        if (!flag || *flag == Host::EOption::Invalid) {
+            WARN("Expected a flag! Got: %s", argv[i]);
+            continue;
+        }
+
+        switch (*flag) {
+        case Host::EOption::Ghost: {
+            ASSERT(i + 1 < argc);
+
+            m_currentGhostFileName = argv[++i];
+            m_currentRawGhost = Abstract::File::Load(m_currentGhostFileName, m_currentRawGhostSize);
+
+            if (m_currentRawGhostSize < System::RKG_HEADER_SIZE ||
+                    m_currentRawGhostSize > System::RKG_UNCOMPRESSED_INPUT_DATA_SECTION_SIZE) {
+                PANIC("File cannot be a ghost! Check the file size.");
+            }
+
+            // Creating the raw ghost file validates it
+            System::RawGhostFile file = System::RawGhostFile(m_currentRawGhost);
+
+            m_currentGhost = new System::GhostFile(file);
+            ASSERT(m_currentGhost);
+        } break;
+        case Host::EOption::Invalid:
+        default:
+            PANIC("Invalid flag!");
+            break;
+        }
+    }
+}
+
+KReplaySystem *KReplaySystem::CreateInstance() {
+    ASSERT(!s_instance);
+    s_instance = new KReplaySystem;
+    return static_cast<KReplaySystem *>(s_instance);
+}
+
+void KReplaySystem::DestroyInstance() {
+    ASSERT(s_instance);
+    auto *instance = s_instance;
```

Comment:
Why not
```c++
delete s_instance;
s_instance = nullptr;
```

Reply:
We use `s_instance` being nullptr as a signal to the heap code that we are actually destroying the instance via `DestroyInstance`, as opposed to it being garbage collected.

Reply:
Oh right